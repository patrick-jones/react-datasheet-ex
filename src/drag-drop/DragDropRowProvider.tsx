import {Component, ReactNode} from 'react';

import { RowDragSource, RowDropTarget } from './row-dnd';
import { InjectedDragDropProps, DragDropRowProps } from './interfaces';


/**
 * Required properties for `DragDropRowProvider`
 */
export interface DragDropRowProviderProps<CellType, V = string>
  extends DragDropRowProps<CellType, V>, InjectedDragDropProps {
  children: (props: InjectedDragDropProps) => ReactNode;
}


class DragDropRow<CellType, V = string>
  extends Component<DragDropRowProviderProps<CellType, V>> {

  render(): ReactNode {
    const {children, ...rest} = this.props;
    return children(rest);
  }
}

/**
 * Provides all drag-and-drop properties to a render function assigned to the
 * `children` property.
 */
const DragDropRowProvider = RowDropTarget(RowDragSource(DragDropRow));

export default DragDropRowProvider;
