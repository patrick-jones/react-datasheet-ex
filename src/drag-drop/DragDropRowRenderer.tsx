import React, {Component} from 'react';

import {InjectedDragDropProps, DragDropRowRendererProps} from './interfaces';
import DragDropRowProvider from './DragDropRowProvider';
import DefaultActionContentRenderer from '../DefaultActionContentRenderer';

const actionCellClassName = 'cell read-only rdr-sheet-renderer__action-cell ' +
  'rdr-sheet-renderer__action-cell--drag-source';

export default class DragDropRowRenderer<CellType, V = string>
  extends Component<DragDropRowRendererProps<CellType, V>> {

  static readonly defaultProps = {
    actionContentRenderer: DefaultActionContentRenderer,
  };

  renderRow = (dndProps: InjectedDragDropProps) => {
    const {connectDragSource, connectDragPreview, connectDropTarget, isDropOver} = dndProps;
    const {row, children, actionContentRenderer} = this.props;
    const ContentRenderer = actionContentRenderer || DefaultActionContentRenderer;

    const dt = isDropOver ? 'rdr-sheet-renderer__row--drop-target' : '';
    const className = `rdr-sheet-renderer__row ${dt}`;
    return connectDropTarget(connectDragPreview(
      <tr key={`$row-${row}`} className={className}>
        {connectDragSource(
          <td key='$$actionCell' className={actionCellClassName}>
            <ContentRenderer row={row} />
          </td>
        )}
        {children}
      </tr>
    ));

  }

  render() {
    const {row, onRowDropped} = this.props;

    return (
      <DragDropRowProvider row={row} onRowDropped={onRowDropped}>
        {this.renderRow}
      </DragDropRowProvider>
    );
  }
}
