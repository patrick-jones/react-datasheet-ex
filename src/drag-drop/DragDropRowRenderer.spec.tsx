import React from 'react';
import {mount} from 'enzyme';

import {DragDropContext} from 'react-dnd';
import Backend, {TestBackend} from 'react-dnd-test-backend';
import toJson from 'enzyme-to-json';

import DragDropRowRenderer from './DragDropRowRenderer';
import DragDropRowProvider from './DragDropRowProvider';


function testableComponent(spyA: jest.Mock, spyB: jest.Mock) {
  return DragDropContext(Backend)(

    class TestContainer extends React.Component {
      render() {
        return (
          <table>
            <tbody>
            <DragDropRowRenderer row={0} cells={[{value: 'Row 0'}]} onRowDropped={spyA}>
              <td>Row 0</td>
            </DragDropRowRenderer>
            <DragDropRowRenderer row={1} cells={[{value: 'Row 1'}]} onRowDropped={spyB}>
              <td>Row 1</td>
            </DragDropRowRenderer>
            </tbody>
          </table>
        );
      }
    }
  );
}


describe('DragDropRowRenderer', () => {
  it('matches snapshot', () => {
    const spyA = jest.fn();
    const spyB = jest.fn();
    const Container = testableComponent(spyA, spyB);

    const wrapper = mount(<Container />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Calls onHeaderDropped', () => {
    const spyA = jest.fn();
    const spyB = jest.fn();
    const Container = testableComponent(spyA, spyB);

    const wrapper = mount(<Container />);
    // console.log(wrapper.debug());
    // @ts-ignore
    const backend: TestBackend = wrapper.instance().getManager().getBackend();

    const hp = wrapper.find(DragDropRowProvider);

    // @ts-ignore
    const dragSource = hp.at(1).childAt(0).instance().getHandlerId();
    // @ts-ignore
    const dropTarget = hp.at(0).instance().getHandlerId();

    backend.simulateBeginDrag([dragSource]);
    backend.simulateHover([dropTarget]);
    backend.simulateDrop();
    backend.simulateEndDrag();

    expect(spyA.mock.calls.length).toBe(1);
    expect(spyB.mock.calls.length).toBe(0);

    expect(spyA.mock.calls[0][0]).toBe(1);
    expect(spyA.mock.calls[0][1]).toBe(0);
  });
});
