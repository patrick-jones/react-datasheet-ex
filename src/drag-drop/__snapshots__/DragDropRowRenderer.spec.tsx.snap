// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DragDropRowRenderer matches snapshot 1`] = `
<DragDropContext(TestContainer)>
  <TestContainer>
    <table>
      <tbody>
        <DragDropRowRenderer
          actionContentRenderer={[Function]}
          cells={
            Array [
              Object {
                "value": "Row 0",
              },
            ]
          }
          onRowDropped={[MockFunction]}
          row={0}
        >
          <DropTarget(DragSource(DragDropRow))
            onRowDropped={[MockFunction]}
            row={0}
          >
            <DragSource(DragDropRow)
              connectDropTarget={[Function]}
              isDropOver={false}
              onRowDropped={[MockFunction]}
              row={0}
            >
              <DragDropRow
                connectDragPreview={[Function]}
                connectDragSource={[Function]}
                connectDropTarget={[Function]}
                isDropOver={false}
                onRowDropped={[MockFunction]}
                row={0}
              >
                <tr
                  className="rdr-sheet-renderer__row "
                  key="$row-0"
                >
                  <td
                    className="cell read-only rdr-sheet-renderer__action-cell rdr-sheet-renderer__action-cell--drag-source"
                    key="$$actionCell"
                  >
                    <DefaultActionContentRenderer
                      row={0}
                    />
                  </td>
                  <td>
                    Row 0
                  </td>
                </tr>
              </DragDropRow>
            </DragSource(DragDropRow)>
          </DropTarget(DragSource(DragDropRow))>
        </DragDropRowRenderer>
        <DragDropRowRenderer
          actionContentRenderer={[Function]}
          cells={
            Array [
              Object {
                "value": "Row 1",
              },
            ]
          }
          onRowDropped={[MockFunction]}
          row={1}
        >
          <DropTarget(DragSource(DragDropRow))
            onRowDropped={[MockFunction]}
            row={1}
          >
            <DragSource(DragDropRow)
              connectDropTarget={[Function]}
              isDropOver={false}
              onRowDropped={[MockFunction]}
              row={1}
            >
              <DragDropRow
                connectDragPreview={[Function]}
                connectDragSource={[Function]}
                connectDropTarget={[Function]}
                isDropOver={false}
                onRowDropped={[MockFunction]}
                row={1}
              >
                <tr
                  className="rdr-sheet-renderer__row "
                  key="$row-1"
                >
                  <td
                    className="cell read-only rdr-sheet-renderer__action-cell rdr-sheet-renderer__action-cell--drag-source"
                    key="$$actionCell"
                  >
                    <DefaultActionContentRenderer
                      row={1}
                    />
                  </td>
                  <td>
                    Row 1
                  </td>
                </tr>
              </DragDropRow>
            </DragSource(DragDropRow)>
          </DropTarget(DragSource(DragDropRow))>
        </DragDropRowRenderer>
      </tbody>
    </table>
  </TestContainer>
</DragDropContext(TestContainer)>
`;
